initialize Jack

thazhemadam opened the pull request.

thazhemadam committed:
build: initialise `plugin.yml`

thazhemadam committed:
feat: initialise `pre-command` and `jack.jl` script

* Set up `jack.jl`, the julia script that performs jack's primary
operations in a self-contained environment in `lib`. This script will
read the values passed as properties to the plugin and appropriately use
JLLPrefixes to apply the specified strategy for the JLLs mentioned.

* Set up `pre-command`, which will instantiate the `lib` environment and
run the `jack.jl` script.

staticfloat commented:
Things to do:

- [x] Allow overriding installation prefix path (default to `/usr/local`)
- [x] Allow overriding copy strategy (`copy`, `symlink`, `hardlink`)
- [x] Emit `--- prefix deploy` section header after Pkg resolution, but before symlink copying so we can properly time the symlinking/copying stage
- [x] Add tests with more detailed package specifications (e.g. name only, as you have here, but also name and version both specificed)
- [ ] Add test that purposefully fails due to invalid specifications (So that we know that we properly surface errors)
- [x] Get rid of extraneous `Pkg.add()`
- [x] Move `BUILDKITE_PLUGIN_JACK_XXX` environment variable parsing into Julia code, so that it's easier to parse both name and version in the Julia code.
- [x] Use a plugin-shipped Manifest to run `JLLPrefixes` without affecting any other environments, similar to https://github.com/staticfloat/sandbox-buildkite-plugin/blob/main/hooks/pre-command#L35

thazhemadam committed:
feat: support specifying JLLs to be installed by name and version

Add support for specifying the JLLs to be installed going by their
name and version.
```
  - JuliaCI/jack:
      - install:
          - name: "<jll-name>"
            version: "<jll-version>"
          - "jll-name"
```

staticfloat commented on lib/jack.jl:
```diff
@@ -0,0 +1,33 @@
+using JLLPrefixes
+
+# Helper to extract buildkite environment arrays
+function extract_env_array(prefix::String)
+    envname(idx::Int) = string(prefix, "_", idx)
+
+    idx = 0
+    array = String[]
+    while haskey(ENV, envname(idx))
+        push!(array, ENV[envname(idx)])
+        idx += 1
+    end
+    return array
+end
```
This is good, but you need to check not only for `BUILDKITE_PLUGIN_JACK_INSTALL_${idx}`, but also `BUILDKITE_PLUGIN_JACK_INSTALL_${idx}_NAME`, since you have the two different syntaxes:

```
            - "FFMPEG_jll"
            - name: "GNUMake_jll"
              version: "v4.4.0+0"
```

staticfloat replied:
For debugging these kinds of things, I like to do this at the top of my plugin hook:

```
echo "--- DEBUGGING ENV DUMP"
env | grep BUILDKITE_PLUGIN_JACK
```

That gives me a nice dump of all the environment variables that I might care about in the log.

thazhemadam replied:
> For debugging these kinds of things, I like to do this at the top of my plugin hook:
> ```
> echo "--- DEBUGGING ENV DUMP"
> env | grep BUILDKITE_PLUGIN_JACK
> ```
> That gives me a nice dump of all the environment variables that I might care about in the log.

Ah, this is nifty. Saves me some time that would've been spent figuring out how other plugins do it otherwise. 😄

thazhemadam commented:
>  Add test that purposefully fails due to invalid specifications (So that we know that we properly surface errors)

Won't this also mean that the buildkite job itself fails? I'm not sure how to catch/test for something like that. 😅

staticfloat commented:
Set the job to soft_fail on exit code 1 (or whatever exit code gets thrown when it fails) then in the command block exit with code 2

staticfloat commented:
I do something similar here: https://github.com/staticfloat/coppermind-buildkite-plugin/blob/main/.buildkite/fail_jobs.yml

thazhemadam committed:
ci: convert pipeline.yml to a higher level pipeline launch tests from it

Instead of `pipeline.yml` running the tests itself, convert it to a
higher level pipeline that can be used to launch other pipelines and
kick-off builds instead; and do the same for launching the tests.

thazhemadam committed:
ci: add tests for failing jobs

Add tests that ensure that jobs expected to fail do so; due to
inappropriate input or Pkg resolutions conflicts due to incompatible
versions.
